Docs contributing (#10062)
apache/incubator-echarts: .github/CONTRIBUTING.md

# Contributing

👍🎉 First off, thanks for taking the time to contribute! 🎉👍

Please check out the [Apache Code of Conduct](https://www.apache.org/foundation/policies/conduct.html).

## Issues

> When opening new issues, please use the [echarts issue helper](https://ecomfe.github.io/echarts-issue-helper/), opening issues in any other way will cause our robot to close them automatically

Reviewer: fyi, I think most projects refer to these as `bot` rather than `robot` :)

> And before you do so, please search for similar questions in [issues list](https://github.com/apache/incubator-echarts/issues?utf8=%E2%9C%93&q=is%3Aissue). If a closed issue have new problems in your environment, please create a new issue and reference the previous one.

Reviewer: This might be better wording:
```
And before doing so, please search for similar questions in our [issues list](https://github.com/apache/incubator-echarts/issues?utf8=%E2%9C%93&q=is%3Aissue). If you are able to reproduce an issue found in a closed issue, please create a new issue and reference the closed one.
```

Please read [documents](http://echarts.apache.org/option.html) carefully before asking any questions.

> Questions in the form on *how to use ECharts to ...* should be at http://stackoverflow.com and will be closed in GitHub issues.

Reviewer 2: ```suggestion
Any questions in the form of *how can I  use echarts to* or *how to use echarts x feature to* belong to [Stack Overflow](http://stackoverflow.com), issues with questions like that in the issue tracker will be closed.
```

> Any questions in the form of *how can I  use echarts to* or *how to use echarts x feature to* belong to [Stack Overflow](http://stackoverflow.com), issues with questions like that in the issue tracker will be closed.

Reviewer: *how can I use echarts to*

Reviewer: It should say `belong in` instead of `belong to`

## Pull Requests

> If you wish to fix a bug or add some new features, please discuss with us in the issue first. If there's no such issue, please create an issue using [ECharts issue helper](https://ecomfe.github.io/echarts-issue-helper/).

Reviewer 2: ```suggestion
If you wish to fix a bug or add new features, please discuss it with us in an issue first. If there's no issue, please create one using the [echarts issue helper](https://ecomfe.github.io/echarts-issue-helper/).
```

## How to Debug ECharts

The following steps help you to set up a developing environment for ECharts.

### 1. Clone ECharts project

If you wish to make pull requests, you should fork the ECharts project first. Otherwise, just clone it locally.

```bash
git clone git@github.com:apache/incubator-echarts.git
```

[ZRender](https://github.com/ecomfe/zrender) is the rendering library under the hood. You need to clone it along with ECharts.

```bash
git clone git@github.com:ecomfe/zrender.git
```

We assume these projects are downloaded at `~/workspace/echarts` and `~/workspace/zrender`. But their locations can be arbitrary.

### 2. Install dependencies

```bash
cd ~/workspace/echarts
npm install
cd ~/workspace/zrender
npm install
```

Sometimes, we need to change the code of ZRender project to fix a problem in ECharts. So we need to make sure we are importing the ZRender in our local project, rather than latest released npm version. To do this, we use an npm feature called `npm link`.

```bash
cd ~/workspace/zrender
npm link
cd ~/workspace/echarts
npm link
```

> And then, you can see `~/workspace/echarts/node_modules/zrender` is a link to `~/workspace/zrender`.

Reviewer 2: ```suggestion
With this, you can see that `~/workspace/echarts/node_modules/zrender` is a link to `~/workspace/zrender`.
```

### 3. Run and debug

To build the ECharts project and watch source file changes (including ZRender project) to rebuild:

```bash
cd ~/workspace/echarts
node build/build.js --watch
```

To build once:

> contributing-inspect.png

Reviewer 3: useless?

Author: @sinchang Thanks. This is a mistake.

```bash
node build/build.js
```

Then, open the test cases under `~/workspace/echarts/test` in Web browser. You can add breakpoints under `src` directory. For example, in Chrome Inspect, it looks like:

![Chrome inspect](../asset/contributing-inspect.png)

### 4. Make a pull request

Fork ECharts project into your own project. Checkout a branch from master branch named `fix-xxxx`, where xxxx is the issue id related. If there's no related issue, you need to create one in most cases to describe what's wrong or what new feature is required.

If you are a committer of apache/incubator-echarts project, which means you have the write access to the project, you still need to push to a new branch (by `git push origin HEAD:refs/heads/fix-xxxx`) and use pull request to push your code. You cannot push code directly to `master` branch, otherwise it will be rejected by GitHub.

> Once your PR is merged, you will be asked to assign a PDF file called [Apache ICLA](https://www.apache.org/licenses/icla.pdf). This is required for all non-trival commits of apache/incubator-echarts project, which means you don't have to sign it if you only changed some spelling in `README.md`.

Reviewer 2: ```suggestion
Once your PR is merged, you will be asked to sign a PDF file called [Apache ICLA](https://www.apache.org/licenses/icla.pdf). This is required for all non-trival commits of the apache/incubator-echarts project, which means you don't have to sign it if you only changed some spelling in `README.md`.
```

> Once your PR is merged, you will be asked to sign a PDF file called [Apache ICLA](https://www.apache.org/licenses/icla.pdf). This is required for all non-trival commits of the apache/incubator-echarts project, which means you don't have to sign it if you only changed some spelling in `README.md`.

Reviewer: Shouldn't the CLA be signed before a PR is merged?

Author: I double checked the regulations, and found that ICLA is required only for new committers (who have write access to the project) rather than all contributors. So this is not necessary in fact.
> A signed ICLA is required to be on file before an individual is given commit rights to any ASF project. https://www.apache.org/licenses/
> If you are a brand new committer, you must first to complete and submit an Individual Contributor License Agreement (ICLA) before your committer account will be activated. https://www.apache.org/dev/new-committers-guide#cla

Reviewer: Gotcha, I noticed this part is removed in #10121 which I think is a good thing but I wonder if we should consider putting this somewhere else so it's clear what the onboarding for new committers (write access to project) is? Or maybe Apache already has one, just wanted to check so it's not forgotten in the future

Author: @ClemMakesApps That's a good idea. I think we should provide this information along with other contribution-related information on our Web site.
